components/constellation/network_listener.rs

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use hyper::header::Location;
+use ipc_channel::ipc;
+use ipc_channel::router::ROUTER;
+use msg::constellation_msg::PipelineId;
+use net_traits::{CoreResourceMsg, FetchMetadata, FetchResponseMsg};
+use net_traits::{IpcSend, NetworkError, ResourceThreads};
+use net_traits::request::RequestInit;
+use net_traits::response::ResponseInit;
+use std::sync::mpsc::Sender;
+
+pub enum NetworkListenerMsg {
+    ResponseMetadata(PipelineId,
+                     (Result<FetchMetadata, NetworkError>,
+                      Vec<Vec<u8>>,
+                      Option<Result<(), NetworkError>>))
+}
+
+pub struct NetworkListener {
+    pub res_init: ResponseInit,
+    pub req_init: RequestInit,
+    pub pipeline_id: PipelineId,
+    pub resource_threads: ResourceThreads,
+    pub sender: Sender<NetworkListenerMsg>,
+    /// chunks which constitute the response body
+    pub chunks: Vec<Vec<u8>>,
+    pub metadata: Option<FetchMetadata>,
+    pub eof: Option<Result<(), NetworkError>>,
+}
+
+
+impl NetworkListener {
+    pub fn initiate_fetch(&self) {
+        let (ipc_sender, ipc_receiver) = ipc::channel().expect("Failed to create IPC channel!");
+
+        let mut listener = NetworkListener {
+            res_init: self.res_init.clone(),
+            req_init: self.req_init.clone(),
+            resource_threads: self.resource_threads.clone(),
+            sender: self.sender.clone(),
+            pipeline_id: self.pipeline_id.clone(),
+            chunks: vec![],
+            metadata: None,
+            eof: None,
+        };
+        ROUTER.add_route(ipc_receiver.to_opaque(), box move |message| {
+            let msg = message.to();
+            match msg {
+                Ok(FetchResponseMsg::ProcessResponse(res)) => listener.check_redirect(res),
+                Ok(FetchResponseMsg::ProcessResponseChunk(chunk)) => listener.chunks.push(chunk),
+                Ok(FetchResponseMsg::ProcessResponseEOF(res_eof)) => {
```

Comment:
It will be better to send the chunks and EOF message to the script thread as they come in, rather than buffering them. This will allow the HTML parser to parse pages like https://html.spec.whatwg.org/ progressively and incrementally, rather than receiving multiple megabytes of content at once.